### ISSUE TITLE
linters.exclusions.warn-unused doesn't catch a same-rule swap under a path exclusion

### ISSUE THREAD

**Issue by yunusdim** [golangci/golangci-lint#6714]
### Welcome

- [x] This is not a request to create or integrate a new linter; those belong in [Discussions](https://github.com/golangci/golangci-lint/discussions/new/choose).
- [x] Yes, I've searched similar [issues on GitHub](https://github.com/golangci/golangci-lint/issues) and didn't find any.
- [x] I agree to follow this project's [Code of Conduct](https://github.com/golangci/golangci-lint?tab=coc-ov-file)

### How did you install golangci-lint?

Official binary

### Your feature request related to a problem? Please describe

An exclusion rule scoped by path (linters.exclusions.rules with a path regex and a linters list) permits every finding of that linter under that path, not a specific finding. warn-unused only checks whether a rule matched anything at all in this run; it doesn't check whether what it matched is the same thing it was matching before.

I reproduced it with golangci-lint 2.12.2. Config: a rule excluding errcheck under legacy/. Starting state: one real errcheck violation in legacy/foo.go (an unchecked error from a call I'll call riskyA), correctly suppressed, and warn-unused correctly silent, the rule is in use. Then I fix riskyA for real (check its error), and add a genuinely new, unrelated unchecked-error call, riskyC, also under legacy/. Running without the config confirms riskyC is a real errcheck violation. Running with the exclusion active: riskyC is suppressed exactly the same way riskyA was, and warn-unused stays silent, because the rule still matches something in that path, just not the same something.

So a real fix and a real regression of the same linter, under the same excluded path, cancel out completely as far as the tool can tell, and the feature built specifically to catch exclusions that no longer apply doesn't fire, because this exclusion does still apply, just to a different violation than the one that justified adding it.

### Describe the solution you'd like

Report, per exclusion rule, how many distinct locations it's currently suppressing, or at least whether that count changed since the rule was added, so a rule that's supposed to be shrinking toward zero but is holding steady or growing is visible.

Short of that, being explicit in the docs that warn-unused answers 'does this rule still match anything' and not 'does this rule still match the same thing it originally covered' would at least set expectations correctly, since right now the exclusions system docs frame these rules mainly as temporary carve-outs for migrating off a linter, which is exactly the case this doesn't protect.

### Describe alternatives you've considered

I don't think per-instance tracking of every matched finding is reasonable to ask for a linter runner this size, that's a bigger design change than the problem probably warrants. A count-based signal, even just a warning if the number of suppressed findings for a rule grows past its original value, would catch the common case without needing full identity tracking.

For context, I ran into this looking at how a few different tools' path/rule exclusion mechanisms handle the same underlying problem, a permitted-violations rule that isn't tied to specific instances. Wrote it up more formally here if useful: https://doi.org/10.5281/zenodo.21908527

### Supporter

- [ ] I am a [sponsor/backer](https://donate.golangci.org) of this project.

**Comment by ldez** [comment]
Technically, the linter is used even if you suppress the reports, so everything is correct here.

You could use the verbose mode to see the number of element excluded:

```console
$ golangci-lint run -v   
...
INFO [runner/exclusion_rules] Skipped 1 issues by rules: [Path: "legacy/", Linters: "errcheck"] 
INFO [runner] Issues before processing: 1, after processing: 0 
INFO [runner] Processors filtering stat (in/out): exclusion_rules: 1/0, cgo: 1/1, filename_unadjuster: 1/1, invalid_issue: 1/1, path_relativity: 1/1, exclusion_paths: 1/1, path_absoluter: 1/1, generated_file_filter: 1/1 
...       
```

**Comment by yunusdim** [comment]
Thanks for looking at it. I think the verbose count answers a different question than the one I was raising.

`Skipped 1 issues by rules: [...]` and `Issues before processing: 1, after processing: 0` are the same in both states I described: before the swap (riskyA suppressed, the violation that justified the rule) and after (riskyA fixed, riskyC suppressed instead). Both print exactly one issue skipped by that rule. So the verbose count is consistent with the rule matching the original violation, or a completely different one, or a third one next month, it can't tell those apart.

The linter is technically used in the sense that it ran and found something. But the exclusion rule was added to cover a specific finding, and by the time riskyA is fixed for real, that finding no longer exists, yet the rule stays silent because something else under the same path now fills the same slot. That's the case warn-unused is meant to catch, a rule that no longer earns its keep, and it can't, because the rule still matches, just not the same thing.

Not asking to reopen if that's not useful to you, just wanted to make sure the actual claim landed, since the count-based argument doesn't address it.
